spatialdata/_core/_spatialdata_ops.py

```diff
@@ -291,154 +289,90 @@ def _describe_paths(paths: list[list[Union[int, str]]]) -> str:
 
-def concatenate(sdatas: list[SpatialData], omit_table: bool = False) -> SpatialData:
-    """Concatenate a list of spatial data objects.
+    if len(set(instance_keys)) == 1:
+        instance_key = list(instance_keys)[0]
+    else:
+        if instance_key is None:
+            raise ValueError("`instance_key` must be specified if tables have different instance keys")
+
+    tables_l = []
+    regions_l = []
+    instance_keys_l = []
+    for k, i, regs, table in zip(region_keys, instance_keys, regions, tables):
+        if isinstance(regs, list):
+            for r in regs:
+                table_ = table[table.obs[k] == r]
+                tables_l.append(table_)
+                regions_l.append(r)
+                instance_keys_l.append(table_.obs[i].values.flatten())
+        else:
+            tables_l.append(table)
+            regions_l.append(regs)
+            instance_keys_l.append(table.obs[i].values.flatten())
+
+    merged_table = anndata.concat(tables_l, label=region_key, keys=regions_l, **kwargs)
+    merged_table.obs[instance_key] = np.concatenate(instance_keys_l, axis=0)
+    attrs = {
+        TableModel.REGION_KEY: regions_l,
+        TableModel.REGION_KEY_KEY: region_key,
+        TableModel.INSTANCE_KEY: instance_key,
+    }
+    merged_table.uns[TableModel.ATTRS_KEY] = attrs
+
+    return TableModel().validate(merged_table)
+
+
+def concatenate(
+    sdatas: list[SpatialData],
+    region_key: str | None = None,
+    instance_key: str | None = None,
+    **kwargs: Any,
+) -> SpatialData:
+    """
+    Concatenate a list of spatial data objects.
 
     Parameters
     ----------
     sdatas
         The spatial data objects to concatenate.
-    omit_table
-        If True, the table is not concatenated. This is useful if the tables are not compatible.
+    region_key
+        The key to use for the region column in the concatenated object.
+        If all region_keys are the same, the `region_key` is used.
+    instance_key
+        The key to use for the instance column in the concatenated object.
+    kwargs
+        See :func:`anndata.concat` for more details.
 
     Returns
     -------
-    SpatialData
-        The concatenated spatial data object.
+    The concatenated :class:`spatialdata.SpatialData` object.
     """
-    from spatialdata._core._spatialdata import SpatialData
+    from spatialdata import SpatialData
 
-    assert type(sdatas) == list
-    assert len(sdatas) > 0
+    assert type(sdatas) == list, "sdatas must be a list"
+    assert len(sdatas) > 0, "sdatas must be a non-empty list"
     if len(sdatas) == 1:
         return sdatas[0]
 
-    if not omit_table:
-        list_of_tables = [sdata.table for sdata in sdatas if sdata.table is not None]
-        merged_table = _concatenate_tables(list_of_tables)
-    else:
-        merged_table = None
+    merged_table = _concatenate_tables(
+        [sdata.table for sdata in sdatas if sdata.table is not None], region_key, instance_key, **kwargs
+    )
```

Comment:
Could the Exeptions bellow this be `KeyError`s or `ValueError`s?

Also these checks are quite cheap, can they happen before the potentially expensive table concatenation?